v2/pkg/protocols/dns/request.go

```diff
@@ -32,6 +32,8 @@ func (request *Request) Type() templateTypes.ProtocolType {
 
 // ExecuteWithResults executes the protocol requests and returns results instead of writing them.
 func (request *Request) ExecuteWithResults(input *contextargs.Context, metadata, previous output.InternalEvent, callback protocols.OutputEventCallback) error {
+	request.options.RateLimiter.Take()
```

Comment:
The rate limiter should be invoked just before sending out the request. If validation fails, we can fail early and proceed to the following item in the queue.